source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,194 @@
+  if (operand_index != 2 && operand_index != 3) {
+    return false;
+  }
+
+  // The new instruction value must exist and must be OpConstant.
+  auto new_instruction = ir_context->get_def_use_mgr()->GetDef(
+      message_.memory_semantics_new_value_id());
+  if (!new_instruction) {
+    return false;
+  }
+  if (new_instruction->opcode() != SpvOpConstant) {
+    return false;
+  }
+  auto new_memory_sematics_value = static_cast<SpvMemorySemanticsMask>(
+      new_instruction->GetSingleWordInOperand(0));
+
+  // Instruction must be atomic instruction only. Operand index must be equal to
+  // 2 in the case of instruction that takes one memory semantics operand.
+  // Instructions that takes two memory semantics, one of the value must be
+  // stronger than the other.
+  switch (needed_atomic_instruction->opcode()) {
+    case SpvOpAtomicLoad:
+    case SpvOpAtomicStore:
+    case SpvOpAtomicExchange:
+    case SpvOpAtomicIIncrement:
+    case SpvOpAtomicIDecrement:
+    case SpvOpAtomicIAdd:
+    case SpvOpAtomicISub:
+    case SpvOpAtomicSMin:
+    case SpvOpAtomicUMin:
+    case SpvOpAtomicSMax:
+    case SpvOpAtomicUMax:
+    case SpvOpAtomicAnd:
+    case SpvOpAtomicOr:
+    case SpvOpAtomicXor:
+    case SpvOpAtomicFlagTestAndSet:
+    case SpvOpAtomicFlagClear:
+    case SpvOpAtomicFAddEXT:
+      if (operand_index != 2) {
+        return false;
+      }
+      break;
+
+    case SpvOpAtomicCompareExchange:
+    case SpvOpAtomicCompareExchangeWeak:
+      if (new_memory_sematics_value ==
+          static_cast<SpvMemorySemanticsMask>(
+              ir_context->get_def_use_mgr()
+                  ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(
+                      operand_index == 3 ? 2 : 3))
+                  ->GetSingleWordInOperand(0))) {
+        return false;
+      }
+      break;
+
+    default:
+      return false;
+  }
+
+  // The new memory semantics value must be larger than old.
+  auto old_memory_sematics_value = static_cast<SpvMemorySemanticsMask>(
+      ir_context->get_def_use_mgr()
+          ->GetDef(
+              needed_atomic_instruction->GetSingleWordInOperand(operand_index))
+          ->GetSingleWordInOperand(0));
+  if (new_memory_sematics_value <= old_memory_sematics_value) {
+    return false;
+  }
+
+  if (!IsValidConverstion(needed_atomic_instruction->opcode(),
+                          new_memory_sematics_value)) {
+    return false;
+  }
+
+  return true;
+}
+
+void TransformationChangingMemorySemantics::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto needed_atomic_instruction =
+      FindInstruction(message_.atomic_instruction(), ir_context);
+  needed_atomic_instruction->SetInOperand(
+      message_.memory_semantics_operand_index(),
+      {message_.memory_semantics_new_value_id()});
+}
+
+bool TransformationChangingMemorySemantics::IsValidConverstion(
+    SpvOp opcode, SpvMemorySemanticsMask new_memory_sematics_value) {
+  switch (opcode) {
```

Comment:
This is not quite right.

Firstly, the old value might have a lot of bits set, other than just the first 5 bits. This is allowed. Thus, we have bits 0-31. We only want to change bits 0-4 (the first 5 bits). 

* Bits 5-31 of the old value must equal bits 5-31 of the new value.

Now, consider just the first 5 bits. I.e. bits 0-4. The logic you have below does not consider the old value, but you need to consider the old value.

E.g. for load:

* if old value is None, then new value must be Acquire or SequentiallyConsistent.
* if old value is Acquire, then new value must be SequentiallyConsistent
* if old value is anything else then return false.

You probably noticed that we can achieve mostly the same thing by doing:

* old value and new value must both be in the set {None, Acquire, SequentiallyConsistent}, and old value must be less than new value.

This is fine, except for barrier instructions; it is not OK to go from acquire -> release. That changes the semantics. You will need to handle barrier instructions more carefully, perhaps just by checking for that one case (acquire -> release) and returning false in that case.

Reply:
Done.